qutebrowser/__init__.py

```diff
@@ -30,4 +30,8 @@
 __version_info__ = tuple(int(part) for part in __version__.split('.'))
 __description__ = "A keyboard-driven, vim-like browser based on Python and Qt."
 
+if 'QTWEBENGINE_DISABLE_SANDBOX' in os.environ:
```

Comment:
Also I think this should probably be done in [earlyinit](https://github.com/qutebrowser/qutebrowser/blob/3ca6910298976937812b5ddf7f500e598f0d971d/qutebrowser/misc/earlyinit.py#L320), not at import time (there the logging module should be set up too). If it could be done later that would be even better. I'm not sure what the actual step it needs to be done before is. Ideally it would be done after we conditionally launch [an existing instance](https://github.com/qutebrowser/qutebrowser/blob/3ca6910298976937812b5ddf7f500e598f0d971d/qutebrowser/app.py#L111). Upstream says to set it before you import any Qt stuff. But is it really? Or is it before you import webengine, or even before you instantiate webengine stuff?

Reply:
Yep, this is definitely the wrong place for it. QtWebEngine [reads the variable](https://github.com/qt/qtwebengine/blob/4a3d409987e4a2dcdcae80310cd478e2b841a1e3/src/core/web_engine_context.cpp#L645) when the `WebEngineContext()` is initialized, which is quite late - here is a quick try with gdb and `--debug`:

```
09:49:57 DEBUG    init       app:init:144 Starting init...
[...]
09:50:22 DEBUG    extensions loader:_load_component:126 Importing qutebrowser.components.adblockcommands
09:50:24 DEBUG    extensions loader:_load_component:126 Importing qutebrowser.components.braveadblock
09:50:24 DEBUG    extensions loader:_load_component:134 Running init hook 'init'
09:50:24 DEBUG    extensions loader:_load_component:126 Importing qutebrowser.components.caretcommands
09:50:24 DEBUG    extensions loader:_load_component:126 Importing qutebrowser.components.hostblock
09:50:24 DEBUG    extensions loader:_load_component:134 Running init hook 'init'
09:50:24 DEBUG    network    hostblock:_should_be_used:83 Configured adblock method auto, adblock library usable: True
09:50:24 DEBUG    extensions loader:_load_component:126 Importing qutebrowser.components.misccommands
09:50:24 DEBUG    extensions loader:_load_component:126 Importing qutebrowser.components.readlinecommands
09:50:24 DEBUG    extensions loader:_load_component:126 Importing qutebrowser.components.scrollcommands
09:50:24 DEBUG    extensions loader:_load_component:126 Importing qutebrowser.components.utils.blockutils
09:50:24 DEBUG    extensions loader:_load_component:126 Importing qutebrowser.components.zoomcommands
09:50:24 DEBUG    init       app:_init_modules:454 Initializing logging from config...
09:50:24 DEBUG    init       log:init_from_config:550 --debug flag overrides log configs
09:50:24 DEBUG    init       app:_init_modules:458 Initializing save manager...
09:50:24 DEBUG    init       app:_init_modules:464 Checking backend requirements...
09:50:24 DEBUG    init       app:_init_modules:467 Initializing prompts...
09:50:24 DEBUG    init       app:_init_modules:470 Initializing network...
09:50:24 DEBUG    init       app:_init_modules:473 Initializing proxy...
09:50:24 DEBUG    init       app:_init_modules:477 Initializing downloads...
09:50:24 DEBUG    init       app:_init_modules:483 Initializing web history...
[...]
09:50:27 DEBUG    init       app:_init_modules:492 Initializing command history...
09:50:27 DEBUG    init       app:_init_modules:495 Initializing websettings...
09:50:27 DEBUG    init       webenginesettings:init:517 Initializing qute://* handler...
09:50:27 DEBUG    init       webenginesettings:init:521 Initializing request interceptor...
09:50:27 DEBUG    init       webenginesettings:init:526 Initializing QtWebEngine downloads...
09:50:27 DEBUG    init       webenginesettings:init:532 Initializing notification presenter...
[...]
09:50:27 DEBUG    init       webenginesettings:init:535 Initializing global settings...
09:50:27 DEBUG    init       webenginesettings:init:539 Initializing profiles...

Thread 1 "python3" hit Breakpoint 1, QtWebEngineCore::WebEngineContext::WebEngineContext (this=<optimized out>, this=<optimized out>) at release/../../../../qtwebengine/src/3rdparty/chromium/base/memory/ref_counted.h:32
32	release/../../../../qtwebengine/src/3rdparty/chromium/base/memory/ref_counted.h: No such file or directory.
```

I think the best place to do those kind of things is actually not earlyinit, but [backendproblem.py](https://github.com/qutebrowser/qutebrowser/blob/qt6-v2/qutebrowser/misc/backendproblem.py). We already have some other backend-specific workarounds (like dealing with stale service worker data) there, and with an appropriate function name (say `_reenable_pyinstaller_macos_sandboxing()`), it'd be much clearer why this is done even without reading the commit message (but agreed, that should at minimum link to the appropriate qutebrowser and PyInstaller issue).

I also agree this should be done *only* when all of those conditions are true:

- Running on Qt 6
- Running on macOS
- Running with PyInstaller (`hasattr(sys, 'frozen')`)

I don't think we can easily write a test for it (tests don't run with PyInstaller) - but if the three conditions above are true, we could actually do this without checking `os.environ` first. Then if it's not needed anymore after a PyInstaller upgrade, the `del` will raise a `KeyError`, and that will be caught by the smoke tests before the release (if not before).